Paper: [WIP] Project Jupyter - Binder 2.0

choldgraf opened the pull request:
Build server here: http://procbuild.scipy.org/

Note: open review goes until July 2nd.

choldgraf committed:
first draft of paper

choldgraf committed:
more authors

choldgraf committed:
Merge pull request #1 from choldgraf/andrewkylejeremey

more authors

betatim committed:
Update paper.rst

choldgraf commented:
@deniederhut any idea why I don't see a build of this PR?

choldgraf committed:
adding sustainability and api section

choldgraf committed:
Merge pull request #3 from choldgraf/extension

adding sustainability and api section

betatim committed:
Update paper.rst

choldgraf committed:
Merge pull request #2 from choldgraf/betatim-patch-1

Update paper with affiliation

willingc committed:
Add email and affiliation

deniederhut commented:
Not sure -- the server is seeing in the test paper when I reopen that PR. I'm going to close and reopen this pull request -- maybe the server was napping when you made the initial PR.

deniederhut commented:
Best guess is that the webhook in procbuild is failing to parse the metadata for this correctly, but I don't have access to that instance. @mpacer do you have time to look into this?

deniederhut commented:
One follow-up -- does it build locally?

choldgraf committed:
Merge pull request #4 from choldgraf/willingc-patch-1

Add email and affiliation

choldgraf commented:
yeah, builds fine for me

deniederhut commented:
The most wonderful M has fixed it! The PR name must start with "Paper" for it to be recognized by the server. You should be able to see your paper here - http://procbuild.scipy.org.

mpacer commented:
Specifically you can find it (for now) here: http://procbuild.scipy.org/download/3

choldgraf commented:
yay! thanks for the clarifications :-)

betatim committed:
Update paper.rst

choldgraf committed:
Merge pull request #6 from choldgraf/betatim-patch-1

Small edits and sustainability changes

jzf2101 committed:
added content but cant fix errors

mpacer committed:
Clarify early intro, introduce terms, add some header info

choldgraf committed:
Merge pull request #8 from mpacer/2018-binder

Clarify early intro, introduce terms, add some header info

mpacer committed:
Rework the cost and sustainability sections

jzf2101 committed:
moved bib and softened language at end

jzf2101 committed:
resolve merge conflict

choldgraf committed:
Merge pull request #7 from jzf2101/2018-binder

wip added lit review but can't figure out rst errors

choldgraf committed:
Merge branch '2018-binder' into 2018-binder

choldgraf committed:
Merge pull request #10 from mpacer/2018-binder

Rework the cost and sustainability sections

choldgraf committed:
updating mybinder.org links and small changes

choldgraf committed:
bugfixes

choldgraf committed:
Merge pull request #11 from choldgraf/final_touches

Final touches

choldgraf commented:
Note that while the content of this paper is finalized, there seem to be some latex rendering problems that I'm not sure how to spot-check. The equation formatting isn't working and the bibliography isn't being inserted properly

http://procbuild.scipy.org/download/3

jzf2101 committed:
added papers mentioned in issue to close #5

choldgraf committed:
Merge pull request #13 from jzf2101/2018-binder

added papers mentioned in issue to close #5

mpacer committed:
fix binderbib file name so bibliography is read in

choldgraf committed:
Merge pull request #14 from mpacer/2018-binder

fix binderbib file name so bibliography is read in

stsievert commented:
I am excited to be reviewing this paper! I'll try to make my review incremental.

choldgraf commented:
great! looking forward to working with you on it @stsievert !

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+
+:author: Jeremy Freeman
+:email:
+:institution:
```
There's a blank footnote because `institution` is blank. I don't know if there's a policy on this. I'd like to see some attribution besides a name. I assume the people with this `institution` tag are open source contributors?

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+for easily creating sharable, interactive, reproducible environments in the
+cloud.
+
+The scientific community is increasingly unified around reproducibility.
```
> The scientific community is increasingly unified around reproducibility.

I'd like a couple citations for this -- without this fact being true, the rest of the paper isn't worth much.

jzf2101 replied:
We've added the following language after that sentence:

>A survey in 2016 of 1,576 researchers reported that 90% of respondents believed
>there exists a reporoducibility crisis in the scientific community.
>A majority of respondents also reported difficulty
>reproducing the work of colleagues :cite:`Baker2016`. Similar results have
>been reported in the cell biology community
>:cite:`The_American_Society_for_Cell_Biology_undated` and machine
>learning community :cite:`Pineau2017-sb`.

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+repositories found that citation of "both visualization tools as well as
+common software packages (such as MATLAB) was a widespread failure" :cite:`Stodden2018-fy`.
+As a result, while reproducibility is *technically* possible, it is not yet
+*practically* possible. This paper introduces Binder 2.0, a tool
```
> As a result, while reproducibility is *technically* possible, it is not yet
*practically* possible.

Maybe "As a result, the technical barriers to reproducibility limit the practical possibility. We introduce Binder 2.0, a tool that we believe makes reproducibility more practical by lowering the technical barriers to sharing computational work."

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+
+.. figure:: images/jupyterlab_binder_ui.png
+   :align: center
+   :figclass: w
```
This is a good image to show; the UI should *definitely* be shown. A couple suggestions:

* should this figure show the shell? It might be useful to show, but only if Binder service supports it (that's one of the features of JupyterLab I've appreciated the most)
* does this colormap have to be used? It looks like it's showing nominal data, which is suited if that's what it's doing. But showing quantitative colormaps is more common; could you use a perceptually uniform colormaps like `viridis`?

jzf2101 replied:
This Binder is the binder of a research publication: https://github.com/dtak/rrr.  We picked this image because it shows how research can be shared on binder.  This colormap is used for a classification task in their experiment.   In this example, the user is exploring the data of the experience in the kernel.  Because JupyterLab is one of the front-ends available on Binder, we had opted to show it because it shows how a user can write live code in addition to running the software in a repo.  Please let us know if you have suggestions to make these distinctions more clear.

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+feeds into the principles we’ve outlined above.
+
+Scalable interactive user sessions (JupyterHub on Kubernetes)
+-------------------------------------------------------------
```
>  Scalable interactive user sessions (JupyterHub on Kubernetes)

Should the title include "(JupyterHub on Kubernetes)"? It references is *how* the section is implemented, not *what* this section is about. Should that information be included in section titles?

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+including Jupyter Notebooks :cite:`Bussonnier2018-kc`, JupyterLab
+:cite:`Project_Jupyter_Contributors_2017-yi`,
+RStudio :cite:`Project_Juptyer_Contributors_2017-ra`, Stencila :cite:`RK_Min_2018-eq`,
+and OpenRefine :cite:`Head2018-jf`.
```
The citations for `Project_Jupyter_Contributors_2017-y`, `Project_Juptyer_Contributors_2017` and `RK_Min_2018-eq` came up as "[?]" in my rendering from Thursday or Friday (June 7th or 8th). These are citations for `JupyterLab`, `RStudio`, and `Stencila`.

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+user intervention. As such, it is straightforward to deploy JupyterHub on any
+major cloud provider.
+
+Kubernetes is designed to be relatively self-healing, and has a declarative
```
> relatively self-healing,

This is ambiguous.

choldgraf replied:
Added in a short extra sentence explaining this a bit more

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+intervention and tinkering.
+
+Finally, because Kubernetes was initially designed to run Google’s web
+services, it is both extremely scalable and battle-tested. A cloud orchestration
```
> Finally, because Kubernetes was initially designed to run Google’s web
services, it is both extremely scalable and battle-tested.

Maybe flip the order and reword? "Finally, Kubernetes is both extremely scalable and battle-tested because the first version ran Google's web services."

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+
+Binder’s solution to this is a lightweight tool called “repo2docker” :cite:`Project_Jupyter_Contributors2017-no`.
+It is an open-source command line tool that converts code repositories into a Docker
+image suitable for running with JupyterHub. Repo2docker does the following things:
```
> Repo2docker does the following things:
> 1. Is called with a single argument, ... .
> 2. Clones the repository, then ....
> 3. Looks for one or more “configuration” .... .
> 4. Constructs a ``Dockerfile`` that builds ...
> 5. 5. Builds an image from ....

Maybe change this so each list item is a complete sentence. Maybe

> Repo2docker 
> 1. is called with a single argument, ... .
> 2. clones the repository, then ....
> 3. looks for one or more “configuration” .... .
> 4. constructs a ``Dockerfile`` that builds ...
> 5. builds an image from ....

Or

> The repo2docker design means that repo2docker
> 1. is called with a single argument, ... .
> 2. clones the repository, then ....
> 3. looks for one or more “configuration” .... .
> 4. constructs a ``Dockerfile`` that builds ...
> 5. builds an image from ....

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+if it exists, triggers the ``get_assemble_scripts`` method, which inserts
+lines into the Dockerfile that install Python and pip.
+
+.. code-block:: python
```
Is this example for `repo2docker`, or Binder? Maybe if it's `repo2docker`s you can take it out.

Either way, some of the lines in this code block spill into the next column, and are either hidden or overrun the text in the other column. This should be fixed

<img width="550" alt="screen shot 2018-06-09 at 2 28 15 pm" src="https://user-images.githubusercontent.com/1320475/41195360-4cd9cbf6-6c1b-11e8-89b8-64eb0b1f32b9.png">

stsievert replied:
> For example, below we show a simplified version of the Python build pack.

I'd expand upon this. Where is this file included in Binder? What does it do?

choldgraf replied:
I've fixed the line overflow - going to leave in the code for now just so we can give readers an idea for the technical structure of the package, but if you feel strongly that it should be left out, we can give it a shot

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+example, if the following ``requirements.txt`` file is present in a repository,
+repo2docker will build an image with Python 3 and the packages pip-installed.
+
+.. code-block:: bash
```
Better syntax for showing files:

``` shell
$ cat requirements.txt
numpy
scipy
matplotlib
```

This goes for the other examples below too (`binder/install.R` and `binder/runtime.txt`)

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+For example, here is the URL for the ``binder-examples`` repository
+that builds a Julia environment:
+
+.. code-block:: bash
```
This also spills into the other column. I think if you do

    For example, here is the URL for the ``binder-examples`` repository
    that builds a Julia environment: ``https://mybinder.org/v2/gh/binder-examples/julia-python/master``

Also, here's a better rewording:

    For example, the URL for the ``binder-examples`` repository
    that builds a Julia environment is ``https://mybinder.org/v2/gh/binder-examples/julia-python/master``

I'm fairly certain you can even have this clickable in the PDF with what ever ReST's link syntax is. I'm pretty sure one thing that would work is

    For example, the URL for the ``binder-examples`` repository
    that builds a Julia environment is
    `https://mybinder.org/v2/gh/binder-examples/julia-python/master <https://mybinder.org/v2/gh/binder-examples/julia-python/master>`

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+    data: {"phase": "ready",
+           "message": "server running at https://hub.mybinder.org/user/<POD-URL>/\n",
+           "url": "https://hub.mybinder.org/user/<POD-URL>/",
+           "token": "<POD-TOKEN>"}
```
Some of this JSON spills of the page.

<img width="473" alt="screen shot 2018-06-09 at 2 41 32 pm" src="https://user-images.githubusercontent.com/1320475/41195452-2775af68-6c1d-11e8-9760-90b6252516c5.png">

choldgraf replied:
arg - ok we'll need to figure out how to reduce the width of all the code cells, thanks for noticing

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+
+There are already several examples of services that use BinderHub’s REST API
+to run webpages and applications that utilize arbitrary kernel execution. For
+example, ``thebelab`` makes it possible to deploy static websites with code blocks
```
Who is ``thebelab``?

choldgraf replied:
thebelab is a project that @minrk is spearheading (https://minrk.github.io/thebelab/) the idea is to have a binder kernel run code that's displayed statically on the page

stsievert replied:
When I read it I was confused. Maybe link and add some words on who they are? e.g.,

> For example, a particular project ([thebelab](http://duckduckgo.com), spearheaded by Min Rk) makes it ...

jzf2101 replied:
We'll include a bibliography citation for clarity.

stsievert commented on papers/project_jupyter/paper.rst:
> ``https://mybinder.org``

If it's linked everywhere else it should be linked here to. 

    `https://mybinder.org <https://mybinder.org>`

(not sure about exact syntax)

stsievert commented on papers/project_jupyter/paper.rst:
Same comment on the titles as with "Scalable interactive user sessions (JupyterHub on Kubernetes)" – I'd like to see "Mybinder.org" removed, especially since it's mentioned in the first section and hyperlinked.

choldgraf replied:
We've changed this section to call it "Cost of running the public Binder service" instead of `mybinder.org`

stsievert commented on papers/project_jupyter/paper.rst:
> By running as a free, public service, we hope that members of the community will find new and interesting applications for Binder. For example, mybinder.org has already been used for reproducible publishing, sharing interactive course materials at the university and high-school level, creating interactive package documentation in Python with Sphinx Gallery, and sharing interactive content that requires a language-specific kernel in order to run.

I think this should be moved up – it shows that Binder is popular and is achieving it's goals!

stsievert replied:
I think I'd like to see the entire section on "Maintaining and sustaining a public service" moved up. It sets the stage for the rest of the paper by detailing that you run a popular online service.

choldgraf replied:
We were debating something like this, but weren't sure if it was appropriate. Our thinking was that the core of the scipy paper is on the BinderHub/repo2docker technology as opposed to the `mybinder.org` service (since ultimately we'd love for other organizations to set up their own BinderHubs). You don't think that it will take away from the tech if we start with a discussion of the service?

stsievert replied:
I think mentioning mybinder first will help keep the reader engaged and interested. Plus it tells a better story and flows better.

choldgraf replied:
I tried moving a short section on `mybinder.org` up to the front of the paper, and leaving the discussions around cost etc at the end. I'm curious to know what you think, and whether this highlights the service well enough.

stsievert commented on papers/project_jupyter/paper.rst:
> one CPU, two GB of RAM.

Change to "one CPU and two GB of RAM."

stsievert commented on papers/project_jupyter/paper.rst:
* What's does the color encoding mean in the "Daily cost by type"?
* Could you rename the `unique_pageviews` and `cost_per_user` to be without underscores and the proper capitalization?
* Could you put xticks on the x-axis for the cost_per_user graph?

choldgraf replied:
done in the upcoming commit!

stsievert commented on papers/project_jupyter/paper.rst:
This math doesn't show up. I think you should use

    :math:`\$220 \times 365 \approx \$80,000`

stsievert replied:
Some of the other math doens't show up too.

choldgraf replied:
yeah we've had a heck of a time figuring out the rST -> latex -> rendered math pipeline :-)

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,740 @@
+If a researcher requires Python 2, 3, RStudio, and Julia, simultaneously for
+their work, repo2docker should enable this.
+
+.. figure:: images/binder_main_ui.png
```
I think this screenshot should go inplace of the RStudio screenshot. The UI is already shown with the Figure 1 and is largely the same as RStudio. This shows something different.

choldgraf replied:
I agree that it's a bit redundant - the reason we'd included it was because many folks assume JupyterHub or Binder is *only* for serving Jupyter Notebooks or Lab, and we wanted to make sure that we gave some love to the non-Jupyter UIs as well (in this case, RStudio). Do you have suggestions for how to highlight the "UIs other than Jupyter Notebook/Lab" outside of having this image?

stsievert replied:
Ah, I didn't express myself clearly. I think they should swap places; I'd like to see this a page or two earlier, in the place where the RStudio screenshot is now.

choldgraf replied:
Ahh I see - I consolidated the RStudio UI to be in the same figure as the JupyterLab UI - will look into moving up the Binder landing page earlier, that's a good point

stsievert reviewed (changes requested):
This looks pretty good. The general structure is good – the features of Binder are outlined before you go get into the details (repo2docker, etc). There are some small points to clear up on this.

The most significant is that I'd like to see the "Maintaining and sustaining a public service" section moved up to the first part of the paper. I think it helps set the stage for the rest of the paper, and let readers know they don't have to have knowledge on servers/Kubernetes/repo2docker.

choldgraf commented:
thanks for the comments @stsievert ! I'll take a pass soon an address the points you've made. see a couple of questions above on some specific comments.

@deniederhut is it OK if we start making edits/responses to reviews now, or should we wait until N >= 1 reviewers have looked through the text?

choldgraf committed:
Merge branch '2018' of https://github.com/scipy-conference/scipy_proceedings into 2018-binder

jzf2101 committed:
added citations for reproducibility crisis, took initial wordig suggestions from review, and started adding institutions

jzf2101 committed:
marked andrewosh as independent

jzf2101 committed:
fixed minrk name spelling

jbednar commented:
I'm trying to get a PDF to look at for review, but the procbuild server just gives me errors:

```
[ 13/06 22:08 Build failed, no pdf can be found ] [WIP] Project Jupyter - Binder 2.0 by choldgraf 
• Checking out proceedings build tools to /procbuild/procbuild/../cache/scipy_proceedings...
Cloning into '/procbuild/procbuild/../cache/scipy_proceedings'...
• Check out paper repository...
Cloning into '/tmp/tmpe8h9h8n6'...
Moving proceedings build tools to temp directory...
• Build the paper...
:46: (WARNING/2) Field list ends without a blank line; unexpected unindent.
Building: project_jupyter
*** Warning: /tmp/tmpe8h9h8n6/publisher/../output/project_jupyter/paper_stats.json does not exist.
PDFLaTeX error output:
================================================================================
b'This is pdfTeX, Version 3.14159265-2.6-1.40.18 (TeX Live 2017/Debian) (preloaded format=pdflatex)\n restricted \\write18 enabled.\nentering extended mode\n(./paper.tex\nLaTeX2e <2017-04-15>\nBabel <3.12> and hyphenation patterns for 3 language(s) loaded.\n(./IEEEtran.cls\nDocument Class: IEEEtran 2015/08/26 V1.8b by Michael Shell\n-- See the "IEEEtran_HOWTO" manual for usage information.\n-- http://www.michaelshell.org/tex/ieeetran/\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/ot1ppl.fd)\n-- Using IEEE Computer Society mode.\n-- Using 8.5in x 11in (letter) paper.\n-- Using PDF output.\n-- This is a 10 point document.\n) (/usr/share/texlive/texmf-dist/tex/latex/cmap/cmap.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/base/ifthen.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/base/fontenc.sty\n(/usr/share/texlive/texmf-dist/tex/latex/base/t1enc.def)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/t1ppl.fd)<>)\n(/usr/share/texlive/texmf-dist/tex/latex/base/inputenc.sty\n(/usr/share/texlive/texmf-dist/tex/latex/base/utf8.def\n(/usr/share/texlive/texmf-dist/tex/latex/base/t1enc.dfu)\n(/usr/share/texlive/texmf-dist/tex/latex/base/ot1enc.dfu)\n(/usr/share/texlive/texmf-dist/tex/latex/base/omsenc.dfu)))\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/color.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics-cfg/color.cfg)\n(/usr/share/texlive/texmf-dist/tex/latex/graphics-def/pdftex.def))\n(/usr/share/texlive/texmf-dist/tex/latex/caption/caption.sty\n(/usr/share/texlive/texmf-dist/tex/latex/caption/caption3.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/keyval.sty))\n\nPackage caption Warning: Unsupported document class (or package) detected,\n(caption) usage of the caption package is not recommended.\nSee the caption package documentation for explanation.\n\n) (/usr/share/texlive/texmf-dist/tex/latex/float/float.sty) (./scipy.sty\n(./status.sty (./draftwatermark.sty (./everypage.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/graphicx.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/graphics.sty\n(/usr/share/texlive/texmf-dist/tex/latex/graphics/trig.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/graphics-cfg/graphics.cfg)))))\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/mathptmx.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/helvet.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/courier.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsmath.sty\nFor additional information on amsmath, use the `?\' option.\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amstext.sty\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsgen.sty))\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsbsy.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsopn.sty))\n(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/amsfonts.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/amsfonts/amssymb.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/tools/bm.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/hyperref.sty\n(/usr/share/texlive/texmf-dist/tex/generic/oberdiek/hobsub-hyperref.sty\n(/usr/share/texlive/texmf-dist/tex/generic/oberdiek/hobsub-generic.sty))\n(/usr/share/texlive/texmf-dist/tex/generic/ifxetex/ifxetex.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/auxhook.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/kvoptions.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/pd1enc.def)\n(/usr/share/texlive/texmf-dist/tex/latex/latexconfig/hyperref.cfg)\n(/usr/share/texlive/texmf-dist/tex/latex/url/url.sty))\n\nPackage hyperref Message: Driver (autodetected): hpdftex.\n\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/hpdftex.def\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/rerunfilecheck.sty))\n\nPackage hyperref Warning: Option `pdftex\' has already been used,\n(hyperref) setting the option has no effect on input line 43.\n\n(/usr/share/texlive/texmf-dist/tex/latex/tools/longtable.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/multirow/multirow.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/base/textcomp.sty\n(/usr/share/texlive/texmf-dist/tex/latex/base/ts1enc.def\n(/usr/share/texlive/texmf-dist/tex/latex/base/ts1enc.dfu)))\n(/usr/share/texlive/texmf-dist/tex/latex/upquote/upquote.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/fancyvrb/fancyvrb.sty\nStyle option: `fancyvrb\' v2.7a, with DG/SPQR fixes, and firstline=lastline fix \n<2008/02/07> (tvz)))\nNo file paper.aux.\n(/usr/share/texlive/texmf-dist/tex/latex/base/ts1cmr.fd)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/t1ptm.fd)\n-- Lines per column: 61 (exact).\n(/usr/share/texlive/texmf-dist/tex/context/base/mkii/supp-pdf.mkii\n[Loading MPS to PDF converter (version 2006.09.02).]\n) (/usr/share/texlive/texmf-dist/tex/latex/oberdiek/epstopdf-base.sty\n(/usr/share/texlive/texmf-dist/tex/latex/oberdiek/grfext.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/latexconfig/epstopdf-sys.cfg))\n(/usr/share/texlive/texmf-dist/tex/latex/caption/ltcaption.sty)\n(/usr/share/texlive/texmf-dist/tex/latex/hyperref/nameref.sty\n(/usr/share/texlive/texmf-dist/tex/generic/oberdiek/gettitlestring.sty))\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/t1phv.fd)\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/ot1ztmcm.fd)<>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/omlztmcm.fd)<>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/omsztmcm.fd)<>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/omxztmcm.fd)<>\n(/usr/share/texlive/texmf-dist/tex/latex/psnfss/ot1ptm.fd)\n\n! LaTeX Error: Counter too large.\n\nSee the LaTeX manual or LaTeX Companion for explanation.\nType H for immediate help.\n ... \n \nl.181 \\renewcommand\n {\\leftmark}{PROC. OF THE 17th PYTHON IN SCIENCE...\n! ==> Fatal error occurred, no output PDF file produced!\nTranscript written on paper.log.\n'
================================================================================
Paper build failed.
A build error occurred during the self.retrieve_pdf step.
```

choldgraf commented:
@jbednar arg, one of the recent commits must have broken it, let me try and resolve this and get back to you

choldgraf committed:
typo in authors

choldgraf commented:
@jzf2101 did the paper build fine for you just before the latest PR? It's not building currently. I fixed a typo that was causing the original error but can't figure out why it isn't currently building...

jzf2101 committed:
started working on additional review comments

jzf2101 committed:
Merge branch '2018-binder' of github.com:choldgraf/scipy_proceedings into 2018-binder

jzf2101 commented:
I believe I know the source of this error.  We have 8 institutions listed on this paper:

- Call Poly SLO
- CZI
- Independent
- Project Juptyer
- Netflix
- Simula Research Lab
- UC Berkeley
- Wild Tree Tech

 and it's generating a ` LaTeX Error: Counter too large.` error.  I believe it breaks after 8 but not at 7. @mpacer I'm going to remove some of the author's institutions for simplicity and we can discuss if there's a smarter way to resolve this issue. 

@jbednar Please try now and let us know if you have issues. Below is the copy of the build I just produced:

[paper.pdf](https://github.com/scipy-conference/scipy_proceedings/files/2103626/paper.pdf)

jzf2101 committed:
paper breaks with 8 institutions but not 7 removing czi for reviewers an will return when we know how to resolve error

choldgraf commented:
thanks for the fix @jzf2101 !

agreiner commented:
Hi folks, I've been assigned to also take a look at this paper. The link from procbuild seems to give a decent PDF now, so I'll be using that, timestamped 20:27 6/14, for my initial review.

agreiner commented:
Wow, great paper. Glad to have the chance to review it! I have a few comments and will also attach a marked-up PDF with proofreading marks.
In the section on "scalable interactive user sessions", second paragraph, you transition to discussing JupyterHub, but at this point you haven't explained the relationship between Binder and JupyterHub. I think you need to add a sentence to do that.

agreiner commented:
In the numbered list under "The BinderHub User Interface", item 6 sort of sounds like the whole session, from item 1 to item 5, is lost when the author logs out. I think you need to clarify what persists to enable others to use the image. In the sentence just after that, you say the URL provided triggers "this process". I think you mean that it triggers the latter half of the process, steps 4-6.

agreiner commented:
In the discussion of the BinderHub REST API, there is an example cited of deploying static web sites with code blocks. It seems odd to describe a web site with interactive code blocks as static. It took me a minute to figure out what you were describing. It's tricky to describe that in such as way that the author of the "static" site (the user of the authoring service) is distinguished from the end user.

agreiner commented:
The Figure 6 caption states that the cloud computing cost per user per day for mybinder.org is 5 cents. The text in paragraph 3 of the "Cost of mybinder.org" section says the hosting cost is 13 cents per user per day. Does the latter include the former? Also, the second panel of that figure needs a color legend.

agreiner commented:
In the section on cost of mybinder.org, you say "Because
Kubernetes is a open source system for managing containers,
while Binder currently runs on the Google Cloud Platform, it
could run on any setup that runs on top of Kubernetes if that
setup proved more cost effective." The causality seems off here. Kubernetes' being open source isn't sufficient to enable Binder to run under any other setup. Also, the phrase 'while Binder currently runs on the GCP' reads like it's meant to be another reason that Binder could run anywhere, which I don't think you intend. I think what you mean is that, while it currently runs on GCP, Binder could run on any setup that runs on top of Kubernetes.

agreiner commented:
Where you discuss the federation model for BinderHub servers, it occurred to me that security would be an issue for hosts of BinderHubs, especially if they run on local hardware. That made me wonder how you deal with sandboxing Binder users.

agreiner commented:
Here is the marked up PDF with the little stuff. This is fantastic work!
[jupyter_binder_review.pdf](https://github.com/scipy-conference/scipy_proceedings/files/2107775/jupyter_binder_review.pdf)

jbednar commented:
To reduce duplicate comments, I reviewed @agreiner's marked-up copy, and added only a couple of my own comments to it (attached).  Looks great, and I agreed with all her edits except for one minor one as noted, in the paper title.  In general I think this is important work described clearly and with clear motivations, so I am happy for it to be part of the proceedings once the minor issues noted are addressed.

[jupyter_binder_review_jbednar.pdf](https://github.com/scipy-conference/scipy_proceedings/files/2112702/jupyter_binder_review_jbednar.pdf)

agreiner commented:
I agree with @jbednar's notes as well. Re the sharable/shareable choice, my only concern here is consistency. (Both spellings occur in the paper currently.) I marked the instances of the least used of the two.

choldgraf commented:
Thanks to @stsievert , @jbednar , and @agreiner for your feedback! We'll take a pass through the text and get back to you ASAP

choldgraf committed:
addressing comments

choldgraf committed:
addressing some reviews

deniederhut commented:
> @deniederhut is it OK if we start making edits/responses to reviews now

Crap dude I'm so sorry I missed this. Yeah, you have the next couple of weeks to make changes to your submission, but you can start as soon as you like.

choldgraf committed:
reviewer comments

choldgraf committed:
Merge pull request #19 from choldgraf/reviews

Responding to reviewer comments

choldgraf committed:
typo

choldgraf commented:
Hey all - I just pushed a bunch of changes to the paper, addressing most of the reviewer comments. Would love anyone (either Binder team or reviewer team) to take a look and add comments / edits / etc! (cc @jzf2101 who's been helping out with the writing on this round)

jbednar commented on papers/project_jupyter/paper.rst:
```diff
@@ -113,17 +113,17 @@ difficult to achieve:
   enabling others to reproduce results without difficulty
 
-Both functional and practical reproducibility depend upon the software and
+Both teachnical and practical reproducibility depend upon the software and
```
technical

choldgraf committed:
clarifying kubernetes hardware

minrk committed:
Author name spelling

and add missing author

minrk committed:
typos

minrk committed:
note EventStream as the protocol used for build messages

minrk committed:
slight rewording of kubernetes requirements

minrk commented:
@choldgraf I added a few small edits and fixed some typos.

choldgraf commented:
thanks @minrk !

stsievert commented on papers/project_jupyter/paper.rst:
```diff
@@ -0,0 +1,780 @@
+
+In the following sections, we describe the three major technical components of
+the Binder project—JupyterHub, repo2docker, and BinderHub—and discuss how each
+feeds into the principles we’ve outlined above.
```
Could you maybe add a sentence saying that these tools were developed with the Jupyter and Binder teams? Kubernetes is featured in the next section and I got confused on who developed repo2docker (and was wondering why an example was shown).

choldgraf replied:
how's https://github.com/choldgraf/scipy_proceedings/pull/20 ?

stsievert replied:
That looks good to me, it helps frame the rest of the paper.

choldgraf committed:
adding note about who developed these tools

jzf2101 committed:
Merge pull request #20 from choldgraf/comment

adding note about who developed these tools

choldgraf commented:
hey all - a friendly ping to @agreiner, @jbednar, and @stsievert to see if there's anything else we should take care of! I think that we addressed all of the comments (or asked clarifying questions etc). 

Here's a link to the latest PDF:

http://procbuild.scipy.org/download/3

If the content looks good then we can spend some time getting the formatting / viz looking proper (e.g. not having two figures on the same page etc)

jbednar commented:
It was in great shape last I looked at it, so I'm happy for it to go in now.

On Wed, Jun 27, 2018, 11:44 Chris Holdgraf <notifications@github.com> wrote:

> hey all - a friendly ping to @agreiner <https://github.com/agreiner> and
> @jbednar <https://github.com/jbednar> to see if there's anything else we
> should take care of! I think that we addressed all of the comments (or
> asked clarifying questions etc).
>
> Here's a link to the latest PDF:
>
> http://procbuild.scipy.org/download/3
>
> If the content looks good then we can spend some time getting the
> formatting / viz looking proper (e.g. not having two figures on the same
> page etc)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy-conference/scipy_proceedings/pull/386#issuecomment-400724107>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABnfCIg-agCQXAeU7Akeol1WRrM7Z5wpks5uA6hEgaJpZM4UD8gS>
> .
>

stsievert commented:
> we can spend some time getting the formatting / viz looking proper

Let me know if you'd like help with this – I can add some comments.

jzf2101 commented:
If you have specific suggestions @stsievert please let us know

agreiner commented:
A few editorial niggles: On page 2, second paragraph, you say "Binder 2.0 utilizes these new tools, and is more scalable...". That should be "and **it** is more scalable...", or you could just go with ", is more scalable..." (depending on how much you want to group those phrases). Near the end of the same paragraph, "frontend" should be "front-end". Where I had suggested adding in the URLs for things, I was thinking you would use references as you did for other URLs. That seems cleaner and still provides the URLs in case someone sees this only in print. This applies to the Binder Site Reliability Guide, Binder Billing, and Binder Grafana Board links referenced on page 6 as well as the links at the top of page 2. Finally, the screen shots in Figure 1 are now awfully small. I think they could go side-by-side and be about 50% larger.
The phrase "On top of" strikes me as ambiguous in the opening sentence of "Maintaining and sustaining a public service".  I think you mean "In addition to providing a showcase for the technical components...".

choldgraf committed:
updating links, small text edits, and moving a figure to avoid clutter

choldgraf committed:
updating UI layout figure

choldgraf committed:
small updates for formatting

choldgraf committed:
Merge pull request #21 from choldgraf/small_edits

Small edits per reviewer comments

choldgraf commented:
Thanks for the suggestions @agreiner - I think I got most of em in https://github.com/choldgraf/scipy_proceedings/pull/21 which I just merged in! LMK if you think there's other stuff to improve!

choldgraf committed:
removing misspelling underline from image

choldgraf committed:
Merge pull request #22 from choldgraf/small_edits

removing misspelling underline from image

agreiner commented:
Looking good! I think the footnotes work for those URLs. Sorry I didn't notice this before, but I was looking closely at Figure 5 and noticed that there is a label for "network", but I don't see a corresponding area in the figure. If it's so small a number as to not register in the graphic, I would leave it out and treat it as part of "miscellaneous". You could just mention the average cost of networking in the caption if it's important.

agreiner commented:
I note you haven't addressed my note on security, but if you feel it's out of scope, I'm fine with leaving that out.

choldgraf commented:
Ah good point, I missed that one re: security @agreiner - on that topic, I'm not sure how to properly address it without using up a lot of space. Perhaps we can link out to some docs on security issues re: kubernetes, since this is a much more general challenge than just Binder

agreiner commented:
FWIW at this point, I'm officially fine with this moving forward once the Fig. 5 image label is addressed.
